EventQueue: Old pointers of sibling were not cleared

deepikabhavnani opened the pull request:
### Description

When adding sibling at the head of linked list, the head if pointing to something in linked list was not updated, hence a loop was formed in linked list

Element0 - First addition to linked list
Element1 - Has higher delay hence added to back
0 ->(next) 1
Element2 - Delay is same as Element0, hence should be sibling of 0, shall be added at head

Expected:
2    ------------->(next) 1
|(sibling)
0

Bug: (Resolved with this)
2    ------------->(next) 1
|(sibling)
0    ------------->(next) 1

If we add more elements and next pointer of sibling is updated, old references will cause issues
Element3 added

Expected:
2    ------------->(next) 3  ------------->(next) 1
|(sibling)
0

Bug: (Resolved with this)
2    ------------->(next) 3  ------------->(next) 1
|(sibling)
0    ------------->(next) 1
***Both siblings here point to different next***

Resolves: https://github.com/ARMmbed/mbed-os/issues/8664

### Pull request type

    [X] Fix
    [ ] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

Commit:
Old pointers of sibling were not cleared

When adding sibling at the head of linked list, the head if pointing
to something in linked list was not updated, hence a loop was formed
in linked list

Element0 - First addition to linked list
Element1 - Has higher delay hence added to back
0 ->(next) 1
Element2 - Delay is same as Element0, hence should be sibling of 0
           Shall be added at head

Expected:
2    ------------->(next) 1
|(sibling)
0

Bug: (Resolved with this)
2    ------------->(next) 1
|(sibling)
0    ------------->(next) 1

If we add more elements and next pointer of sibling is updated, old
references will cause issues
Element3 added

Expected:
2    ------------->(next) 3  ------------->(next) 1
|(sibling)
0

Bug: (Resolved with this)
2    ------------->(next) 3  ------------->(next) 1
|(sibling)
0    ------------->(next) 1
***Both siblings here point to different next***

cmonr reviewed (approved):
Makes sense to me.

pan- commented:
@deepikabhavnani Would it be possible to extend the event unit tests to make sure the problem is solved ?

deepikabhavnani commented:
> @deepikabhavnani Would it be possible to extend the event unit tests to make sure the problem is solved ?

Yes. Wanted to confirm first if next should be `null` in case of siblings, or all of them point to same next and update on `head` should traverse and update `next` of all siblings. Not sure how it was designed to be. @geky

Commit:
Corrected destructor loop to clear all pending events

In `equeue_destroy` the external loop was for main events linked
list and internal loop for siblings.
Siblings start was not initialized correctly for each main link

deepikabhavnani commented:
Additional commit to correct traverse in destructor

Commit:
Added test case for verification

0xc0170 reviewed (approved):
Bugfix with the test 💯

cmonr commented:
CI started

cmonr commented:
Waiting on a final review by @geky, but otherwise should be good to go.

geky commented:
Sorry for such a late review, this looks like a perfect fix :+1: 

Looking into it, the equeue relies on sibling's next pointers being NULL here:
https://github.com/ARMmbed/mbed-os/blob/master/events/equeue/equeue.c#L294

Unfortunately it was a hard catch because it is only an issue when canceling siblings which _also_ have a sibling. In this case the equeue tries to update the next event's reference when sibling is not NULL, which is wrong and corrupts the data structure. The is then made worse because events references can end up unused in a lot of use cases, making the corruption go unnoticed.

![image](https://user-images.githubusercontent.com/1075160/50352814-fe419200-050b-11e9-80ef-dcce86ab9236.png)

Thanks for the fix! After looking at the code again I'm embarrassed the equeue has been used for this long without some sort of design doc, sorry you guys had to reverse engineer all of that.

cmonr commented:
~~CI started~~

My mistake. Starting CI has become a reflex now. Updated status accordingly.
